Replace usage of legacy Mbed TLS crypto Kconfigs with PSA Crypto API

valeriosetti opened the pull request:
Just replace usage of legacy Kconfigs related to Mbed TLS crypto with corresponding PSA Crypto API.

ceolin commented on subsys/net/lib/sockets/Kconfig:
```diff
@@ -131,12 +131,9 @@ config NET_SOCKETS_SOCKOPT_TLS
-	imply MBEDTLS_PKCS1_V21 if !NET_L2_OPENTHREAD
+	imply PSA_WANT_ALG_RSA_PKCS1V15_SIGN if !NET_L2_OPENTHREAD
+	imply PSA_WANT_ALG_RSA_PSS if !NET_L2_OPENTHREAD
 	imply MBEDTLS_KEY_EXCHANGE_RSA_ENABLED if !NET_L2_OPENTHREAD
```
Is this still available in MBEDTLS 4.0 ?

valeriosetti replied:
Mmmm replacing `MBEDTLS_KEY_EXCHANGE_RSA_ENABLED` with some other key exchange seems to create a lot of other problems in various samples. If you don't mind I will create a specific PR for this.

valeriosetti committed:
modules: mbedtls: add ciphersuite specific Kconfigs

Add new Kconfigs which are specific for certain TLS ciphersuite. The idea
is that instead of enabling key-exchanges related Kconfigs and then adding
all possible dependencies/selections there (which is not strictly correct)
the user enables a specific ciphersuite for which is easier and more
accurate to specify dependencies.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
net: lib: sockets: remove selection of crypto features from socket TLS

There is no reason for which CONFIG_NET_SOCKETS_SOCKOPT_TLS should select
some crypto features by default. The correct list really depends on the
server/client the device would like to connect to.
For this reason the selection of any crypto feature is removed from
CONFIG_NET_SOCKETS_SOCKOPT_TLS with the idea that the proper ciphersuite
is to be enabled in the final sample/test configuration file.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
samples: net|mgmt|thrift: sockets: fix crypto and TLS Kconfigs

Use the TLS ciphersuite Kconfigs to select the desired TLS connection mode
instead of manually selecting all the crypto/key-exchanges Kconfigs.

For 'samples/subsys/mgmt/hawkbit' use the 'overlay-tls.conf' file for
testing 'sample.net.hawkbit.tls' instead of repeating the Kconfigs in the
yaml file.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
tests: net|thrift: fix crypto and TLS Kconfigs

Use the TLS ciphersuite Kconfigs to select the desired TLS connection mode
instead of manually selecting all the crypto/key-exchanges Kconfigs.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
tests: net: socket: tls_configurations: remove unnecessary Kconfigs

Now that CONFIG_NET_SOCKETS_SOCKOPT_TLS doesn't select unnecessary crypto
features there is no need to re-disable them in the main configuration
file of this test code.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
modules: hostap: use PSA Crypto API by default

Set CONFIG_WIFI_NM_WPA_SUPPLICANT_CRYPTO_MBEDTLS_PSA enabled by default
meaning that PSA Crypto API are used by default instead of legacy crypto.
Some Kconfigs for EC and RSA keys are also enabled there in order to
allow support of private key management.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
tests: net: lib: coap_server: simplify test cases

'net.coap.server.common' and 'net.coap.server.secure' test the very same
thing. Both include 'prj.conf' file which configure the TLS connection,
but then 'net.coap.server.secure' re-add these TLS Kconfigs also in the
yaml file which is pointless.

Since we're talking about a secure connection in both cases, remove
'net.coap.server.common' and keep only 'net.coap.server.secure'. Then
for the latter use only 'prj.conf' to hold the configuration.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
test: net: coap_server: adjust main stack size

Increase stack size to prevent runtime crashes on testing.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
doc: migration-guide: add notes for changes in socket TLS and HostAP

* Socket TLS: crypto features are no more automatically selected and this
  must be explicitly done for each application.
* HostAP: usage of PSA Crypto API is enabled by default.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

tomi-font commented on modules/mbedtls/Kconfig.ciphersuites:
Why use `imply` and not `select`? If you are enabling a ciphersuite, you want all the Kconfig options to be unconditionally enabled, no?

tomi-font commented on modules/mbedtls/Kconfig.ciphersuites:
```diff
@@ -0,0 +1,87 @@
+
+config MBEDTLS_CIPHERSUITE_TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256
+	bool "Ciphersuite TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256"
+	imply MBEDTLS_SSL_PROTO_TLS1_2
```
Are all the ciphersuites in this file available only with TLS 1.2 (not 1.3)?

tomi-font commented on doc/releases/migration-guide-4.4.rst:
```diff
@@ -633,6 +633,10 @@ Networking
 
+* Automatic selection of crypto Kconfigs has been removed from :kconfig:option:`NET_SOCKETS_SOCKOPT_TLS`
+  as these strongly depends on the final application's needs. As a consequence desired TLS protocol
+  version and ciphersuite(s) must be explicitly selected.
```
Advertise the new helpers? using https://github.com/zephyrproject-rtos/zephyr/pull/90681

tomi-font commented on doc/releases/migration-guide-4.4.rst:
```diff
@@ -633,6 +633,10 @@ Networking
   response transmission.
 
+* Automatic selection of crypto Kconfigs has been removed from :kconfig:option:`NET_SOCKETS_SOCKOPT_TLS`
+  as these strongly depends on the final application's needs. As a consequence desired TLS protocol
```
```suggestion
  as they strongly depend on the final application's needs. As a consequence the desired TLS protocol
```

valeriosetti commented on modules/mbedtls/Kconfig.ciphersuites:
```diff
@@ -0,0 +1,87 @@
+	imply PSA_WANT_ALG_TLS12_PRF
+	imply PSA_WANT_KEY_TYPE_DERIVE
+
+config MBEDTLS_CIPHERSUITE_TLS_RSA_WITH_AES_256_CBC_SHA256
```
IDK, I would prefer not to create some other intermediate Kconfig that wraps too many things (like what's currently happening with key exchanges Kconfigs).

valeriosetti commented on west.yml:
```diff
@@ -326,7 +328,8 @@ manifest:
     - name: mcuboot
-      revision: 9ac72969f281491d677e669d053281fc2d538ed4
+      remote: mcuboot-upstream
+      revision: 31fcf095a0c22e14c143b28e909c85f9309157ae
```
IIRC the last time I faced a similar situation I was told that Zephyr's fork of McuBoot is periodically synced with the upstream version and that no PR should be opened on Zephyr's fork. @de-nordic Am I right? If this is the case, can we bump the version now so that I can include the commit I need for this PR?
